File: solutions/javascript/LetterCombinationsofaPhoneNumber/LetterCombinationsofaPhoneNumber.js
var res_arr = [];
var arr = {'2':'abc', '3':'def', '4':'ghi', '5':'jkl', '6':'mno', '7':'pqrs', '8':'tuv', '9':'wxyz'};
/**
 * @param {string} digits
 * @return {string[]}
 */
var letterCombinations = function(digits) {
    if (digits == '') {
        return [];
    }
    res_arr = [];
    gars(digits, 0, digits.length, '');
    return res_arr;
};
//get all result string
var gars = function(digits, idx, n, res) {
    if (idx < n) {
        var s = arr[digits[idx]];
        for (var i = 0; i < s.length; i++) {
            gars(digits, idx + 1, n, res + s[i]);
        }
    } else {
        res_arr.push(res);
    }
};

File: solutions/java/LetterCombinationsofaPhoneNumber/LetterCombinationsofaPhoneNumber.java
class Solution {
    private String[] arr = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};

    public List<String> letterCombinations(String digits) {
        List<String> result = new ArrayList<>();
        if (digits.isEmpty()) {
            return result;
        }
        return getResult(result, digits, 0, "");
    }

    private List<String> getResult(List<String> result, String digits, int idx, String res) {
        if (idx == digits.length()) {
            result.add(res);
            return result;
        }
        String tmp = arr[digits.charAt(idx) - '0'];
        if (tmp.isEmpty()) {
            return getResult(result, digits, idx + 1, res);
        }
        for (int i = 0; i < tmp.length(); i++) {
            result = getResult(result, digits, idx + 1, res + tmp.charAt(i));
        }
        return result;
    }
}